Configure TLS in the Kaniko Chains demo

zregvart opened the pull request:
When using exported OpenShift registry, i.e. on CRC
`default-route-openshift-image-registry.apps-crc.testing`, we need to
configure `kaniko` to trust the ingress CA certificate.
This is done here by creating a configmap `ingress-ca.crt` with the
ingress CA certificate and merging it into the Tekton Chains secret, so
it can accessed when mounted in the pod running the `kaniko` task.
This allows us to use `--registry-certificate` parameter to point to it.

Ref. https://issues.redhat.com/browse/HACBS-105

simonbaird commented on hack/chains/kaniko-demo.sh:
```diff
@@ -21,17 +21,29 @@ oc project tekton-chains
+  oc create configmap ingress-ca.crt --from-file=ingress-ca.crt=<(echo $INGRESS_CA) 2>&1 | grep -q 'already exists' \
+    && oc get configmap ingress-ca.crt -o yaml | yq eval ".data.\"ingress-ca.crt\" = \"$INGRESS_CA\"" - | oc replace -f - # or in case it already exists replace it
+  # merge the configmap created above into $DOCKERCFG secret used by tekton-chains-controller and the task below
+  yq eval-all 'select(fileIndex == 0) * select(fileIndex == 1)' <(oc get configmap ingress-ca.crt -o yaml) <(oc get secret tekton-chains-controller-token-wdmvh -o yaml) | oc replace -f -
```
The  name `tekton-chains-controller-token-wdmvh` makes me wonder if it needs to be looked up in some other way. Would the the `wdmvh` part be different in different instances?

zregvart replied:
Oh darn, left that hardcoded should be `$DOCKERCFG`

simonbaird commented on hack/chains/kaniko-demo.sh:
```diff
@@ -21,17 +21,29 @@ oc project tekton-chains
+  # Fetch the ingress CA certificate
+  INGRESS_CA=$(oc get secret router-ca -n openshift-ingress-operator -o jsonpath="{.data.tls\.crt}" )
+  # create configmap with the ingress CA certificate
+  oc create configmap ingress-ca.crt --from-file=ingress-ca.crt=<(echo $INGRESS_CA) 2>&1 | grep -q 'already exists' \
```
It might not matter, but if $INGRESS_CA could have line breaks in it, then perhaps it's better to put quotes in `echo "$INGRESS_CA"`

simonbaird commented on hack/chains/kaniko-demo.sh:
```diff
@@ -21,17 +21,29 @@ oc project tekton-chains
+  INGRESS_CA=$(oc get secret router-ca -n openshift-ingress-operator -o jsonpath="{.data.tls\.crt}" )
+  # create configmap with the ingress CA certificate
+  oc create configmap ingress-ca.crt --from-file=ingress-ca.crt=<(echo $INGRESS_CA) 2>&1 | grep -q 'already exists' \
+    && oc get configmap ingress-ca.crt -o yaml | yq eval ".data.\"ingress-ca.crt\" = \"$INGRESS_CA\"" - | oc replace -f - # or in case it already exists replace it
```
It's a nice create or replace trick I haven't seen before.. :)

simonbaird commented:
Looks good to me. As far as I can tell it doesn't clash with anything in #6 but we should ask @caugello to confirm he agrees with that.

Thinking about the best place for this config: For something demo-specific, I think setting this up "just-in time" in the demo script is fine (versus in the Gitops hook). In the long term we're not going to care about the demos any more, so it's better not to have demo specific setup in Gitops I think.

On the other hand, actually I'm not 100% clear on whether the extra config make being created here is generally useful to provide a generic solution for accessing the internal registry with TLS. If it is, then perhaps the generally useful part could be extracted and incorporated into the gitops hook. But I'm also not clear about how it compares to what's already being done in #6, -- I hope you and @caugello have a clear picture and are in agreement.

simonbaird commented:
Lgtm.

simonbaird commented:
Needs a revision after rebasing on #6 IIUC .

zregvart committed:
Configure TLS in the Kaniko Chains demo

When using exported OpenShift registry, i.e. on CRC
`default-route-openshift-image-registry.apps-crc.testing`, we need to
configure `kaniko` to trust the ingress CA certificate.

In order to do that the chains-ca-cert secret created in #6 is used and
mounted to the `kaniko-chains` taks using a volume. This diverges from
the default `kaniko-chains` task so we need to patch it with the volume
and volumeMount.

Given that the key holding the trusted certificates in the
`chains-ca-cert` secret is `ca-certificates.crt` this matches one of the
paths that golang would looks at when loading trusted CA certificates
(`/etc/ssl/certs/ca-certificates.crt`), and this makes it nicely
suitable for the mount point.

Ref. https://issues.redhat.com/browse/HACBS-105

caugello commented:
lgtm

simonbaird commented:
Lgtm. You could consider squashing and updating the commit message accordingly, but I don't mind either way.

zregvart commented:
Squashed and commit message reworded.
